gmso/tests/test_topology.py

```diff
@@ -352,7 +354,29 @@ def test_dihedral_dihedraltype_update(self):
         assert len(top.dihedral_types) == 1
         assert len(top.dihedral_type_expressions) == 1
         assert len(top.atom_type_expressions) == 2
-        
+
+    def test_improper_impropertype_update(self):
+        top = Topology()
+
+        atype1 = AtomType(expression='sigma + epsilon')
+        atype2 = AtomType(expression='sigma * epsilon')
+        site1 = Site('a', atom_type=atype1)
+        site2 = Site('b', atom_type=atype2)
+        site3 = Site('c', atom_type=atype2)
+        site4 = Site('d', atom_type=atype1)
+        atype = ImproperType()
+        improper = Improper(connection_members=[site1, site2, site3, site4], connection_type=atype)
+        top.add_site(site1)
+        top.add_site(site2)
+        top.add_site(site3)
+        top.add_site(site4)
+        top.add_connection(improper)
+
+        assert top.n_impropers == 1
+        assert len(top.improper_types) == 1
+        assert len(top.improper_type_expressions) == 1
+        assert len(top.atom_type_expressions) == 2
```

Comment:
What would this return? Should they be 0s or 1s?
```python3
        assert top.n_dihedrals == 0
        assert len(top.dihedral_types) == 0
        assert len(top.dihedral_type_expressions) ==0
```

General question - should query-like properties lump impropers in with dihedrals? And does this mean we should have `.n_dihedral`, `.n_proper`, and `.n_improper`?

Reply:
I would argue for a complete separation of proper and improper dihedrals. I am open to other ideas, but I lean towards calling proper dihedrals 'dihedrals' and improper dihedrals 'impropers'. This would be consistent with lammps, which has `dihedral_style` and `improper style`. It would be consistent with Cassandra, which has `dihedrals` and `impropers` as different sections of the MCF file. It would be inconsistent with GROMACS, which has `proper dihedrals` and `improper dihedrals`.

Reply:
This fluid representation is the concern I have. I agree that propers and impropers should be separated, I'm just not sure how to do the bookkeepping. Is "dihedral" an umbrella term for everything (propers and impropers) or does it imply propers only? It seems like GROMACS thinks about it as the latter, and LAMMPS & Cassandra the former. My current preference is for the GROMACS way of doing things, although I can be convinced otherwise (and internal consistency matters more than perfect coherence to other libraries)

Reply:
Completely agree that consistency and clarity is most important. 

I can be talked into `.n_dihedral = .n_proper + .n_improper`. Curious what others think?

Reply:
> Completely agree that consistency and clarity is most important.
> 
> I can be talked into `.n_dihedral = .n_proper + .n_improper`. Curious what others think?

That seems like a decent way to handle it. Keep the proper and impropers separate like we expect, then can write them out however the writer needs it.

At least it fits my mental model of proper and improper dihedrals.

Reply:
I also agree with that, but I think we should rename `dihedral` in another PR. This one is already pretty big